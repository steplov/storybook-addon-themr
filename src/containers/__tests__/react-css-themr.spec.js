/* eslint-disable class-methods-use-this,react/prop-types */
import React from 'react';
import { shallow } from 'enzyme';
import ReactCssThemr from '../react-css-themr';
import { EVENT_ID_INIT, EVENT_ID_DATA } from '../../index';

jest.mock('react-css-themr', () => ({
  ThemeProvider: ({ children }) => (
    <div>
      ThemeProvider
      {children}
    </div>
  )
}));

const noop = () => {};
const story = () => (
  <div>Story</div>
);
const themes = {
  'theme 1': {
    a: 12,
    b: 13
  },
  theme2: {
    c: 14,
    d: 15
  },
  4: {
    e: 22,
    f: 33
  }
};

describe('<ReactCssThemr />', () => {
  it('should render empty without props', () => {
    const wrapper = shallow(
      <ReactCssThemr
        themes={themes}
        channel={{
          emit: noop,
          on: noop,
          removeListener: noop
        }}
        story={story()}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it(`should subcribe on ${EVENT_ID_DATA} event`, () => {
    class MockedReactCssThemr extends ReactCssThemr {
      onInitChannel() {}
    }

    const on = jest.fn();

    MockedReactCssThemr.prototype.onReceiveData = jest.fn().mockReturnValue(42);

    shallow(
      <MockedReactCssThemr
        themes={themes}
        channel={{
          emit: noop,
          on,
          removeListener: noop
        }}
        story={story()}
      />
    );

    expect(on).toHaveBeenCalled();
    expect(on.mock.calls[0][0]).toBe(EVENT_ID_DATA);
    expect(on.mock.calls[0][1]()).toBe(42);
  });

  it(`should emit ${EVENT_ID_INIT} event with current theme and themes`, () => {
    const emit = jest.fn();

    shallow(
      <ReactCssThemr
        themes={themes}
        channel={{
          emit,
          on: noop,
          removeListener: noop
        }}
        story={story()}
      />
    );

    expect(emit).toHaveBeenCalled();
    expect(emit.mock.calls[0][0]).toBe(EVENT_ID_INIT);
    expect(emit.mock.calls[0][1].themes).toEqual(themes);
    expect(emit.mock.calls[0][1].currentTheme).toBeTruthy();
  });

  it(`should unsubcribe from ${EVENT_ID_INIT} event`, () => {
    const removeListener = jest.fn();
    const onReceiveData = jest.fn();
    const wrapper = shallow(
      <ReactCssThemr
        themes={themes}
        channel={{
          emit: noop,
          on: noop,
          removeListener
        }}
        story={story()}
      />
    );

    wrapper.instance().onReceiveData = onReceiveData;
    wrapper.unmount();

    expect(removeListener).toHaveBeenCalledWith(EVENT_ID_DATA, onReceiveData);
  });

  describe('onReceiveData', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(
        <ReactCssThemr
          themes={themes}
          channel={{
            emit: noop,
            on: noop,
            removeListener: noop
          }}
          story={story()}
        />
      );
    });

    it('should set current theme from query to state if it exists', () => {
      wrapper.instance().onReceiveData({ currentTheme: 'theme 1' });

      expect(wrapper.state('currentTheme')).toBe('theme 1');
    });

    it('should mark that component is ready to show data', () => {
      wrapper.instance().onReceiveData({ currentTheme: 'theme 1' });
      expect(wrapper.state('isReady')).toBe(true);
    });
  });

  it('renders story if addon cannot be applied', () => {
    const wrapper = shallow(
      <ReactCssThemr
        themes={themes}
        channel={{
          emit: noop,
          on: noop,
          removeListener: noop
        }}
        story={story()}
      />
    );

    expect(wrapper.html()).toBe('<div>Story</div>');
  });

  it('renders wrapped story', () => {
    const wrapper = shallow(
      <ReactCssThemr
        themes={themes}
        channel={{
          emit: noop,
          on: noop,
          removeListener: noop
        }}
        story={story()}
      />
    );

    wrapper.setState({
      isReady: true
    });

    expect(wrapper.html()).toBe('<div>ThemeProvider<div>Story</div></div>');
  });
});
