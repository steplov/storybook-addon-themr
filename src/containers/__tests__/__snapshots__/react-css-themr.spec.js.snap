// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReactCssThemr /> should render empty without props 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    Story
</div>,
  "nodes": Array [
    <div>
      Story
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ReactCSSThemr
        channel={
                Object {
                        "emit": [Function],
                        "on": [Function],
                        "removeListener": [Function],
                      }
        }
        story={
                <div>
                        Story
                </div>
        }
        themes={
                Object {
                        "4": Object {
                          "e": 22,
                          "f": 33,
                        },
                        "theme 1": Object {
                          "a": 12,
                          "b": 13,
                        },
                        "theme2": Object {
                          "c": 14,
                          "d": 15,
                        },
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ReactCSSThemr {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onReceiveData": [Function],
        "props": Object {
          "channel": Object {
            "emit": [Function],
            "on": [Function],
            "removeListener": [Function],
          },
          "story": <div>
            Story
</div>,
          "themes": Object {
            "4": Object {
              "e": 22,
              "f": 33,
            },
            "theme 1": Object {
              "a": 12,
              "b": 13,
            },
            "theme2": Object {
              "c": 14,
              "d": 15,
            },
          },
        },
        "refs": Object {},
        "state": Object {
          "currentTheme": "4",
          "isReady": false,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          Story
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          Story
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ReactCSSThemr
    channel={
        Object {
            "emit": [Function],
            "on": [Function],
            "removeListener": [Function],
          }
    }
    story={
        <div>
            Story
        </div>
    }
    themes={
        Object {
            "4": Object {
              "e": 22,
              "f": 33,
            },
            "theme 1": Object {
              "a": 12,
              "b": 13,
            },
            "theme2": Object {
              "c": 14,
              "d": 15,
            },
          }
    }
/>,
}
`;
